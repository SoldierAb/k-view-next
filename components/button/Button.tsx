import { computed, defineComponent, toRefs } from "vue"
import buttonTypes from "./buttonTypes"
import { useLocaleReceive } from "../config-provider"
import useConfigInject from "../_hooks/useConfigInject"
import { componentNamePrefix } from "../_constants/root"

export default defineComponent({
  name: componentNamePrefix + "Button",
  props: buttonTypes(),
  emits: ["click"],
  setup(props, { attrs, emit }) {
    // 多语言注入
    const locale = useLocaleReceive("Button")
    const handleClick = (ev: Event) => {
      if (attrs.disabled) {
        ev.preventDefault()
        return
      }
      emit("click", ev)
    }
    const { prefixCls } = useConfigInject()
    const { type } = toRefs(props)
    const classes = computed(() => {
      const pre = prefixCls.value
      return [`${pre}-btn-container`, `${pre}-btn-${type.value||'default'}`,attrs.class]
    })
    const btnProps = {
      ...attrs,
      class: classes.value,
      onClick: handleClick,
    }
    return {
      btnProps,
      locale,
      type
    }
  },
  render() {
    const { locale, btnProps, $slots, type } = this
    const child = $slots.default?.()
    return (
      type === 'confirm'?
      <button {...btnProps}>
      {locale.OkText}
    </button>:
      <button {...btnProps}>
        {child}
      </button>
    )
  },
})
