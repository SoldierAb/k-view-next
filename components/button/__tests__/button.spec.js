import { mount } from '@vue/test-utils'
import Button from '../index'

describe('Button', () => {
    it('renders correctly', () => {
        const wrapper = mount({
            render() {
                return <Button>Follow</Button>
            },
        })
        expect(wrapper.html()).toMatchSnapshot()
    })

    it('click event should be called correctly', () => {
        const fn = jest.fn()
        const wrapper = mount({
            render (){
                return <Button onClick={fn}>点击</Button>
            },
        })
        wrapper.trigger('click')
        expect(fn).toBeCalled()
        wrapper.trigger('click')
        expect(fn).toBeCalledTimes(2)
    })
})