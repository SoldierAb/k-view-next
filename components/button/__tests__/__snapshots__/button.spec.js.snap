// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button renders correctly 1`] = `<button class="k-btn-container k-btn-default">Follow</button>`;
